common/environment/environment.cpp

```diff
@@ -2615,6 +2614,44 @@ void getRoxieProcessServers(const char *process, SocketEndpointArray &servers)
     getRoxieProcessServers(queryRoxieProcessTree(root, process), servers);
 }
 
+extern ENVIRONMENT_API RoxieTargetType readRoxieTargetType(const char *roxieName)
+{
+    RoxieTargetType roxieTargetType = RTTUnknown;
+    if (isEmptyString(roxieName))
+        return roxieTargetType;
+
+    bool hasPortZero = false, hasNonZeroPort = false;
+    Owned<IEnvironmentFactory> factory = getEnvironmentFactory(true);
+    Owned<IConstEnvironment> environment = factory->openEnvironment();
+    Owned<IPropertyTree> envRoot = &environment->getPTree();
+
+    VStringBuffer xpath("Software/RoxieCluster[@name='%s'][1]", roxieName);
+    IPropertyTree *cluster = envRoot->queryPropTree(xpath.str());
+    if (!cluster)
+        return roxieTargetType;
+    Owned<IPropertyTreeIterator> farmers = cluster->getElements("RoxieFarmProcess");
+    ForEach(*farmers)
+    {
+        IPropertyTree &farmer = farmers->query();
+        const char *port = farmer.queryProp("@port");
+        if (!port)
+            continue;
+        if (streq(port, "0"))
+            hasPortZero = true;
```

Comment:
suggestion: A hint for how to simplify code like this (but not for this PR).  It could be coded as:

        if (streq(port, "0"))
            roxieTargetType |= RTTQueued;
        else
            roxieTargetType |= RTTPublished;
        if (roxieTargetType == RTTBoth)
            break;
    }

and avoid the code at the end of the function.